quil-rs/src/expression/simplification/mod.rs

```diff
@@ -0,0 +1,153 @@
+use crate::expression::Expression;
+
+mod by_hand;
+
+/// Simplify an [`Expression`].
+pub(super) fn run(expression: &Expression) -> Expression {
+    by_hand::run(expression)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::str::FromStr;
+
+    macro_rules! test_simplify {
+        ($name:ident, $input:expr, $expected:expr) => {
+            #[test]
+            fn $name() {
+                let parsed_input = Expression::from_str($input);
+                assert!(parsed_input.is_ok(), "Parsing input `{}` failed!", $input);
+                let parsed_expected = Expression::from_str($expected);
+                assert!(
+                    parsed_expected.is_ok(),
+                    "Parsing expected expression `{}` failed!",
+                    $expected
+                );
+                let computed = run(&parsed_input.unwrap());
+                assert_eq!(parsed_expected.unwrap(), computed);
```

Comment:
A bit more idiomatic:

```suggestion
                let parsed_input = Expression::from_str($input).unwrap_or_else(|error| {
                    panic!("Parsing input `{}` failed!", $input)
                });
                let parsed_expected = Expression::from_str($expected).unwrap_or_else(|error| {
                    panic!("Parsing expected expression `{}` failed!", $expected)
                });;
                let computed = run(&parsed_input);
                assert_eq!(parsed_expected, computed);
```

Reply:
👍, going with a slight modification of this to use the `|error|` in the message.